EIPS/eip-1559.md

```diff
@@ -9,130 +9,107 @@ category: Core
+* `BASE_FEE_MAX_CHANGE`: `CURRENT_BLOCK.parent.BASE_FEE / BASE_FEE_MAX_CHANGE_DENOMINATOR`. The maximum amount the `BASE_FEE` can change between blocks (either up or down).
+* `INITIAL_BASE_FEE` : 0.000000001 ETH (1 nanoeth)
+* `BASE_FEE_CHANGE`: `CURRENT_BLOCK.parent.BASE_FEE`. The amount that the BASE_FEE will increase in `CURRENT_BLOCK` relative to `CURRENT_BLOCK.parent`.
+
+### Process
+* At `block.number == INITIAL_FORK_BLOCK_NUMBER` we set `BASE_FEE = INITIAL_BASE_FEE`
+* `BASE_FEE` is set as follows
   * Let `delta = block.gas_used - TARGET_GASUSED` (possibly negative).
-  * Set `BASEFEE = PARENT_BASEFEE + PARENT_BASEFEE * delta // TARGET_GASUSED // BASEFEE_MAX_CHANGE_DENOMINATOR`
-  * Clamp the resulting `BASEFEE` inside of the allowable bounds if needed, where a valid `BASEFEE` is one such that `abs(BASEFEE - PARENT_BASEFEE) <= max(1, PARENT_BASEFEE // BASEFEE_MAX_CHANGE_DENOMINATOR)`
+  * Set `BASE_FEE = PARENT_BASE_FEE + PARENT_BASE_FEE * delta // TARGET_GASUSED // BASE_FEE_MAX_CHANGE_DENOMINATOR`
+  * Clamp the resulting `BASE_FEE` inside of the allowable bounds if needed, where a valid `BASE_FEE` is one such that `abs(BASE_FEE - PARENT_BASE_FEE) <= max(1, PARENT_BASE_FEE // BASE_FEE_MAX_CHANGE_DENOMINATOR)`
 * EIP-1559 transactions are encoded as `rlp([nonce, gasLimit, to, value, data, gasPremium, feeCap, v, r, s])` where `v,r,s` is a signature of `rlp([nonce, gasLimit, to, value, data, gasPremium, feeCap])`.
 * During the transition phase, legacy transactions (`rlp([nonce, gasPrice, gasLimit, to, value, data, v, r, s])`) will continue to be valid and can be included as part of the `LEGACY_GAS_LIMIT` block of transactions.
 * To produce an EIP1559 transactions, the new `GAS_PREMIUM` and `FEECAP` fields are set as follows:
   * `GAS_PREMIUM` serves as a "tip" to the miner
   * `FEECAP` serves as the absolute maximum that the transaction sender is willing to pay
 * During transaction execution, for EIP1559 transactions we calculate the cost to the `tx.origin` and the gain to the `block.coinbase` as follows:
-  * Set `GASPRICE = min(BASEFEE + tx.GasPremium, tx.fee_cap)`
+  * Set `GASPRICE = min(BASE_FEE + tx.GasPremium, tx.fee_cap)`
   * Let `GASUSED` be the gas used during the transaction execution/state transition
   * The `tx.origin` initially pays `GASPRICE * tx.gasLimit`, and gets refunded `GASPRICE * (tx.gasLimit - GASUSED)`
-  * The `block.coinbase` gains `(GASPRICE - BASEFEE) * GASUSED`.
-    * If `GASPRICE < BASEFEE` (due to the `FEECAP`), this means that the `block.coinbase` _loses_ funds from this operation; in this case, we check that the post-balance is non-negative and throw an exception if it is negative.
+  * The `block.coinbase` gains `(GASPRICE - BASE_FEE) * GASUSED`.
+    * If `GASPRICE < BASE_FEE` (due to the `FEECAP`), this means that the `block.coinbase` _loses_ funds from this operation; in this case, we check that the post-balance is non-negative and throw an exception if it is negative.
```

Comment:
This should be removed and instead we assert gas_price >= base_fee

Reply:
Hmm, current implementations do not follow this rule?  I'm not a big fan of it so would be happy to delete it, but I left it in because my goal was to not change the spec at all in this PR.

Reply:
Specifically, the specification does *NOT* assert that `fee_cap >= base_fee`.  This was intentional in the original design, but I would support removing it, especially if implementations currently do not implement this per spec.

Reply:
I would need to dig up or ask @vbuterin for the specifics but there was a very good reason to make this change- miners can potentially set the coinbase address to someone else's and burn the funds by this mechanism.

The geth implementation does assert that gas_price >= base_fee, and I believe the Besu one does as well as they have been tested to work together (although, perhaps not under those conditions).

Reply:
We definitely should generate a test case that covers that condition.  😄  Once I get the spec converted to Python, I have a personal task to create some tests (which would pass in the python spec code).  Though, that may come *after* I get 2718 tests generated so could be a while.

Reply:
I would propose that we merge this as-is, and we have a separate PR to add the assertion.  This is primarily because I dislike mixing normative and non-normative changes in a PR, especially when the non-normative changes are as sweeping as the ones here.

Reply:
Great point, let's go ahead and merge this. I'm willing to bet Besu already has that test condition I just don't know for certain.